File: pi42/order.py
"""
Order API endpoints for Pi42.
"""

from typing import Dict, Any, Optional, List, Union


class OrderAPI:
    """
    Order API client for managing Pi42 orders.
    """

    def __init__(self, client):
        """
        Initialize the Order API client.

        Args:
            client: The main Pi42Client instance
        """
        self._client = client

    def place_order(self, symbol: str, side: str, order_type: str, quantity: float,
                   place_type: str = "ORDER_FORM", margin_asset: str = "INR",
                   price: Optional[float] = None, reduce_only: bool = False,
                   take_profit_price: Optional[float] = None,
                   stop_loss_price: Optional[float] = None,
                   stop_price: Optional[float] = None,
                   position_id: Optional[str] = None,
                   device_type: str = "WEB",
                   user_category: str = "EXTERNAL") -> Dict[str, Any]:
        """
        Place an order on Pi42's trading platform.

        Args:
            symbol: Trading pair symbol (e.g., 'BTCUSDT')
            side: Order side ('BUY' or 'SELL')
            order_type: Order type ('MARKET', 'LIMIT', 'STOP_MARKET', 'STOP_LIMIT')
            quantity: Amount of the asset to be ordered
            place_type: Type of order placement ('ORDER_FORM' or 'POSITION')
            margin_asset: Asset used for margin (e.g., 'INR')
            price: Price for LIMIT orders
            reduce_only: Whether the order should only reduce the position
            take_profit_price: Price at which take profit order should be executed
            stop_loss_price: Price at which stop loss order should be executed
            stop_price: Required for STOP_MARKET and STOP_LIMIT orders
            position_id: Position ID (required if placeType is 'POSITION')
            device_type: Device type ('WEB', 'MOBILE')
            user_category: User category ('EXTERNAL', 'INTERNAL')

        Returns:
            Dict: Order placement response
        """
        endpoint = "/v1/order/place-order"
        
        params = {
            "placeType": place_type,
            "quantity": quantity,
            "side": side,
            "symbol": symbol,
            "type": order_type,
            "reduceOnly": reduce_only,
            "marginAsset": margin_asset,
            "deviceType": device_type,
            "userCategory": user_category
        }
        
        # Add conditional parameters
        if price is not None:
            params["price"] = price
            
        if take_profit_price is not None:
            params["takeProfitPrice"] = take_profit_price
            
        if stop_loss_price is not None:
            params["stopLossPrice"] = stop_loss_price
            
        if stop_price is not None:
            params["stopPrice"] = stop_price
            
        if position_id is not None:
            params["positionId"] = position_id
            
        return self._client.post_request(endpoint, params)

    def add_margin(self, position_id: str, amount: Union[int, float]) -> Dict[str, Any]:
        """
        Add margin to a specific position.

        Args:
            position_id: Unique identifier for the position
            amount: Amount of margin to be added

        Returns:
            Dict: Response containing margin addition details
        """
        endpoint = "/v1/order/add-margin"
        
        params = {
            "positionId": position_id,
            "amount": amount
        }
        
        return self._client.post_request(endpoint, params)

    def reduce_margin(self, position_id: str, amount: Union[int, float]) -> Dict[str, Any]:
        """
        Reduce the margin on an existing trading position.

        Args:
            position_id: Unique identifier for the position
            amount: Amount of margin to reduce

        Returns:
            Dict: Response containing margin reduction details
        """
        endpoint = "/v1/order/reduce-margin"
        
        params = {
            "positionId": position_id,
            "amount": amount
        }
        
        return self._client.post_request(endpoint, params)

    def get_open_orders(self, page_size: Optional[int] = None, 
                       sort_order: Optional[str] = None,
                       start_timestamp: Optional[int] = None,
                       end_timestamp: Optional[int] = None,
                       symbol: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Retrieve open orders for the account.

        Args:
            page_size: Number of results to return per page
            sort_order: Sorting order ('asc' or 'desc')
            start_timestamp: Start timestamp for filtering open orders
            end_timestamp: End timestamp for filtering open orders
            symbol: Trading pair symbol to filter orders

        Returns:
            List[Dict]: List of open orders
        """
        endpoint = "/v1/order/open-orders"
        
        params = {}
        
        if page_size:
            params["pageSize"] = page_size
        if sort_order:
            params["sortOrder"] = sort_order
        if start_timestamp:
            params["startTimestamp"] = start_timestamp
        if end_timestamp:
            params["endTimestamp"] = end_timestamp
        if symbol:
            params["symbol"] = symbol
            
        return self._client.get_request(endpoint, params)

    def get_order_history(self, page_size: Optional[int] = None, 
                         sort_order: Optional[str] = None,
                         start_timestamp: Optional[int] = None,
                         end_timestamp: Optional[int] = None,
                         symbol: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Retrieve historical order data for the account.

        Args:
            page_size: Number of results to return per page
            sort_order: Sorting order ('asc' or 'desc')
            start_timestamp: Start timestamp for filtering order history
            end_timestamp: End timestamp for filtering order history
            symbol: Trading pair symbol to filter orders

        Returns:
            List[Dict]: List of historical orders
        """
        endpoint = "/v1/order/order-history"
        
        params = {}
        
        if page_size:
            params["pageSize"] = page_size
        if sort_order:
            params["sortOrder"] = sort_order
        if start_timestamp:
            params["startTimestamp"] = start_timestamp
        if end_timestamp:
            params["endTimestamp"] = end_timestamp
        if symbol:
            params["symbol"] = symbol
            
        return self._client.get_request(endpoint, params)

    def get_linked_orders(self, link_id: str) -> List[Dict[str, Any]]:
        """
        Retrieve orders that are linked by a specific link ID.

        Args:
            link_id: The unique identifier for the linked orders

        Returns:
            List[Dict]: List of linked orders
        """
        endpoint = f"/v1/order/linked-orders/{link_id}"
        
        return self._client.get_request(endpoint)

    def fetch_margin_history(self, symbol: Optional[str] = None,
                            page_size: Optional[int] = None, 
                            sort_order: Optional[str] = None,
                            start_timestamp: Optional[int] = None,
                            end_timestamp: Optional[int] = None) -> Dict[str, Any]:
        """
        Retrieve the margin history for an account.

        Args:
            symbol: Trading pair symbol to filter margin history
            page_size: Number of results to return per page
            sort_order: Sorting order ('asc' or 'desc')
            start_timestamp: Start timestamp for filtering margin history
            end_timestamp: End timestamp for filtering margin history

        Returns:
            Dict: Margin history response with data and pagination info
        """
        endpoint = "/v1/order/fetch-margin-history"
        
        params = {}
        
        if symbol:
            params["symbol"] = symbol
        if page_size:
            params["pageSize"] = page_size
        if sort_order:
            params["sortOrder"] = sort_order
        if start_timestamp:
            params["startTimestamp"] = start_timestamp
        if end_timestamp:
            params["endTimestamp"] = end_timestamp
            
        return self._client.get_request(endpoint, params)

    def delete_order(self, client_order_id: str) -> Dict[str, Any]:
        """
        Delete a specific order based on its client order ID.

        Args:
            client_order_id: The unique identifier for the order to be deleted

        Returns:
            Dict: Order deletion response
        """
        endpoint = "/v1/order/delete-order"
        
        params = {
            "clientOrderId": client_order_id
        }
        
        return self._client.delete_request(endpoint, params)

    def cancel_all_orders(self) -> Dict[str, Any]:
        """
        Cancel all open orders.

        Returns:
            Dict: Order cancellation response
        """
        endpoint = "/v1/order/cancel-all-orders"
        
        return self._client.delete_request(endpoint)

File: order.go
package pi42

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"strconv"
)

// OrderAPI provides access to order management endpoints
type OrderAPI struct {
	client *Client
}

// NewOrderAPI creates a new Order API instance
func NewOrderAPI(client *Client) *OrderAPI {
	return &OrderAPI{client: client}
}

// PlaceOrderParams represents parameters for placing an order
type PlaceOrderParams struct {
	Symbol          string    `json:"symbol"`
	Side            OrderSide `json:"side"`
	Type            OrderType `json:"type"`
	Quantity        float64   `json:"quantity"`
	PlaceType       string    `json:"placeType"`
	MarginAsset     string    `json:"marginAsset"`
	Price           float64   `json:"price,omitempty"`
	ReduceOnly      bool      `json:"reduceOnly"`
	TakeProfitPrice float64   `json:"takeProfitPrice,omitempty"`
	StopLossPrice   float64   `json:"stopLossPrice,omitempty"`
	StopPrice       float64   `json:"stopPrice,omitempty"`
	PositionID      string    `json:"positionId,omitempty"`
	DeviceType      string    `json:"deviceType"`
	UserCategory    string    `json:"userCategory"`
	Leverage        int       `json:"leverage,omitempty"`
}

// OrderResponse represents the structured response when placing an order
type OrderResponse struct {
	ClientOrderID       string  `json:"clientOrderId"`
	Time                string  `json:"time"`
	Symbol              string  `json:"symbol"`
	ContractType        string  `json:"contractType"`
	Type                string  `json:"type"`
	Side                string  `json:"side"`
	Price               float64 `json:"price"`
	OrderAmount         float64 `json:"orderAmount"`
	FilledAmount        float64 `json:"filledAmount"`
	AvailableBalance    float64 `json:"availableBalance"`
	LinkID              string  `json:"linkId"`
	LinkType            string  `json:"linkType"`
	SubType             string  `json:"subType"`
	PlaceType           string  `json:"placeType"`
	LockedMargin        float64 `json:"lockedMargin"`
	BaseAsset           string  `json:"baseAsset"`
	QuoteAsset          string  `json:"quoteAsset"`
	MarginAsset         string  `json:"marginAsset"`
	LockedMarginInAsset float64 `json:"lockedMarginInMarginAsset"`
	Leverage            int     `json:"leverage"`
	ID                  float64 `json:"id"`
	StopPrice           float64 `json:"stopPrice"`
}

// PlaceOrder places an order on Pi42's trading platform
func (api *OrderAPI) PlaceOrder(params PlaceOrderParams) (OrderResponse, error) {
	endpoint := "/v1/order/place-order"

	// Convert struct to map for the request
	paramsMap := map[string]interface{}{
		"symbol":      params.Symbol,
		"side":        params.Side,
		"type":        params.Type,
		"quantity":    params.Quantity,
		"reduceOnly":  params.ReduceOnly,
		"marginAsset": params.MarginAsset,
	}

	// Add other parameters if they're set
	if params.PlaceType != "" {
		paramsMap["placeType"] = params.PlaceType
	} else {
		paramsMap["placeType"] = "ORDER_FORM"
	}

	if params.Price != 0 {
		paramsMap["price"] = params.Price
	}

	if params.TakeProfitPrice != 0 {
		paramsMap["takeProfitPrice"] = params.TakeProfitPrice
	}

	if params.StopLossPrice != 0 {
		paramsMap["stopLossPrice"] = params.StopLossPrice
	}

	if params.StopPrice != 0 {
		paramsMap["stopPrice"] = params.StopPrice
	}

	if params.PositionID != "" {
		paramsMap["positionId"] = params.PositionID
		paramsMap["placeType"] = "POSITION"
	} else {
		paramsMap["placeType"] = "ORDER_FORM"
	}

	if params.Leverage > 0 {
		paramsMap["leverage"] = params.Leverage
	}

	data, err := api.client.Post(endpoint, paramsMap, false)
	if err != nil {
		return OrderResponse{}, err
	}

	var result OrderResponse
	if err := json.Unmarshal(data, &result); err != nil {
		return OrderResponse{}, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// AddMargin adds margin to a specific position
func (api *OrderAPI) AddMargin(positionID string, amount float64) (map[string]interface{}, error) {
	endpoint := "/v1/order/add-margin"

	params := map[string]interface{}{
		"positionId": positionID,
		"amount":     amount,
	}

	data, err := api.client.Post(endpoint, params, false)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// ReduceMargin reduces the margin on an existing trading position
func (api *OrderAPI) ReduceMargin(positionID string, amount float64) (map[string]interface{}, error) {
	endpoint := "/v1/order/reduce-margin"

	params := map[string]interface{}{
		"positionId": positionID,
		"amount":     amount,
	}

	data, err := api.client.Post(endpoint, params, false)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// OrderQueryParams represents parameters for querying orders
type OrderQueryParams struct {
	PageSize       int    `json:"pageSize,omitempty"`
	SortOrder      string `json:"sortOrder,omitempty"`
	StartTimestamp int64  `json:"startTimestamp,omitempty"`
	EndTimestamp   int64  `json:"endTimestamp,omitempty"`
	Symbol         string `json:"symbol,omitempty"`
}

// GetOpenOrders retrieves open orders for the account with structured response
func (api *OrderAPI) GetOpenOrders(params OrderQueryParams) ([]OpenOrder, error) {
	endpoint := "/v1/order/open-orders"

	queryParams := make(map[string]string)

	if params.PageSize > 0 {
		queryParams["pageSize"] = strconv.Itoa(params.PageSize)
	}
	if params.SortOrder != "" {
		queryParams["sortOrder"] = params.SortOrder
	}
	if params.StartTimestamp > 0 {
		queryParams["startTimestamp"] = strconv.FormatInt(params.StartTimestamp, 10)
	}
	if params.EndTimestamp > 0 {
		queryParams["endTimestamp"] = strconv.FormatInt(params.EndTimestamp, 10)
	}
	if params.Symbol != "" {
		queryParams["symbol"] = params.Symbol
	}

	data, err := api.client.Get(endpoint, queryParams, false)
	if err != nil {
		return nil, err
	}

	var result []OpenOrder
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// GetOrderHistory retrieves historical order data with structured response
func (api *OrderAPI) GetOrderHistory(params OrderQueryParams) ([]OrderHistoryItem, error) {
	endpoint := "/v1/order/order-history"

	queryParams := make(map[string]string)

	if params.PageSize > 0 {
		queryParams["pageSize"] = strconv.Itoa(params.PageSize)
	}
	if params.SortOrder != "" {
		queryParams["sortOrder"] = params.SortOrder
	}
	if params.StartTimestamp > 0 {
		queryParams["startTimestamp"] = strconv.FormatInt(params.StartTimestamp, 10)
	}
	if params.EndTimestamp > 0 {
		queryParams["endTimestamp"] = strconv.FormatInt(params.EndTimestamp, 10)
	}
	if params.Symbol != "" {
		queryParams["symbol"] = params.Symbol
	}

	data, err := api.client.Get(endpoint, queryParams, false)
	if err != nil {
		return nil, err
	}

	var result []OrderHistoryItem
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// GetLinkedOrders retrieves orders that are linked by a specific link ID
func (api *OrderAPI) GetLinkedOrders(linkID string) ([]LinkedOrder, error) {
	endpoint := fmt.Sprintf("/v1/order/linked-orders/%s", linkID)

	data, err := api.client.Get(endpoint, nil, false)
	if err != nil {
		return nil, err
	}

	var result []LinkedOrder
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// FetchMarginHistory retrieves the margin history for an account
func (api *OrderAPI) FetchMarginHistory(params OrderQueryParams) (map[string]interface{}, error) {
	endpoint := "/v1/order/fetch-margin-history"

	queryParams := make(map[string]string)

	if params.PageSize > 0 {
		queryParams["pageSize"] = strconv.Itoa(params.PageSize)
	}
	if params.SortOrder != "" {
		queryParams["sortOrder"] = params.SortOrder
	}
	if params.StartTimestamp > 0 {
		queryParams["startTimestamp"] = strconv.FormatInt(params.StartTimestamp, 10)
	}
	if params.EndTimestamp > 0 {
		queryParams["endTimestamp"] = strconv.FormatInt(params.EndTimestamp, 10)
	}
	if params.Symbol != "" {
		queryParams["symbol"] = params.Symbol
	}

	data, err := api.client.Get(endpoint, queryParams, false)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return result, nil
}

// DeleteOrder deletes a specific order based on its client order ID
func (api *OrderAPI) DeleteOrder(clientOrderID string) (*OrderCancelResponse, error) {
	endpoint := "/v1/order/delete-order"

	params := map[string]interface{}{
		"clientOrderId": clientOrderID,
	}

	data, err := api.client.Delete(endpoint, params)
	if err != nil {
		return nil, err
	}

	var result OrderCancelResponse
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return &result, nil
}

// CancelAllOrders cancels all open orders with structured response
func (api *OrderAPI) CancelAllOrders() (*BatchCancelResponse, error) {
	endpoint := "/v1/order/cancel-all-orders"

	data, err := api.client.Delete(endpoint, map[string]interface{}{})
	if err != nil {
		return nil, err
	}

	var result BatchCancelResponse
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	return &result, nil
}

// BulletParams represents simplified parameters for quick order placement
type BulletParams struct {
	Symbol     string    // Trading pair symbol
	Side       OrderSide // BUY or SELL
	OrderType  OrderType // MARKET, LIMIT, STOP_MARKET, or STOP_LIMIT
	Price      float64   // Required for LIMIT and STOP_LIMIT orders
	StopPrice  float64   // Required for STOP_MARKET and STOP_LIMIT orders
	Count      float64   // Multiplier for minimum quantity
	ReduceOnly bool      // Whether this is a reduce-only order
	Leverage   int       // Leverage to use for the order (optional)
	PositionID string    // Position ID for the order (optional)
}

// Bullet creates an order using exchange specifications for precision and minimum quantity
// and returns a structured order response
func (api *OrderAPI) Bullet(params BulletParams) (*OrderResponse, error) {
	// Get contract info for the symbol
	contractInfo, ok := api.client.ExchangeInfo[params.Symbol]
	if !ok {
		return nil, fmt.Errorf("symbol %s not found in exchange info", params.Symbol)
	}

	// Validate order type
	validOrderTypes := []OrderType{"MARKET", "LIMIT", "STOP_MARKET", "STOP_LIMIT"}
	isValidType := false
	for _, orderType := range validOrderTypes {
		if params.OrderType == orderType {
			isValidType = true
			break
		}
	}

	if !isValidType {
		return nil, fmt.Errorf("invalid order type: %s. Must be MARKET, LIMIT, STOP_MARKET, or STOP_LIMIT", params.OrderType)
	}

	// Validate required parameters for specific order types
	if (params.OrderType == "LIMIT" || params.OrderType == "STOP_LIMIT") && params.Price <= 0 {
		return nil, fmt.Errorf("price must be specified and greater than 0 for %s orders", params.OrderType)
	}

	if (params.OrderType == "STOP_MARKET" || params.OrderType == "STOP_LIMIT") && params.StopPrice <= 0 {
		return nil, fmt.Errorf("stopPrice must be specified and greater than 0 for %s orders", params.OrderType)
	}

	// Determine the minimum quantity based on order type
	var minQuantity float64
	var maxQuantity float64

	if params.OrderType == "MARKET" || params.OrderType == "STOP_MARKET" {
		minQuantity = contractInfo.MarketMinQuantity
		maxQuantity = contractInfo.MarketMaxQuantity
	} else {
		minQuantity = contractInfo.MinQuantity
		maxQuantity = contractInfo.MaxQuantity
	}

	// If min quantity is not set (could happen if filter parsing failed), use a safe default
	if minQuantity <= 0 {
		log.Default().Printf("Warning: Minimum quantity for %s not set, using default value\n", params.Symbol)
		minQuantity = 0.001 // Default fallback
	}

	// Calculate quantity based on minimum quantity and count
	quantity := minQuantity * params.Count

	// Check if quantity exceeds the maximum
	if maxQuantity > 0 && quantity > maxQuantity {
		return nil, fmt.Errorf("calculated quantity %.8f exceeds maximum allowed %.8f for %s",
			quantity, maxQuantity, params.Symbol)
	}

	// Round to the correct precision
	quantity = roundToDecimal(quantity, contractInfo.QuantityPrecision)

	// Check if the order type is supported for this symbol
	// For stop orders, we check if the base type (MARKET/LIMIT) is supported
	baseOrderType := params.OrderType
	if params.OrderType == "STOP_MARKET" {
		baseOrderType = "MARKET"
	} else if params.OrderType == "STOP_LIMIT" {
		baseOrderType = "LIMIT"
	}

	orderTypeSupported := false
	for _, supportedType := range contractInfo.OrderTypes {
		if supportedType == baseOrderType {
			orderTypeSupported = true
			break
		}
	}

	if !orderTypeSupported {
		return nil, fmt.Errorf("order type %s not supported for symbol %s",
			baseOrderType, params.Symbol)
	}

	// Determine default margin asset based on contract info
	marginAsset := contractInfo.QuoteAsset
	if len(contractInfo.MarginAssets) > 0 {
		marginAsset = contractInfo.MarginAssets[0]
	}

	// Set up order parameters
	orderParams := PlaceOrderParams{
		Symbol:      params.Symbol,
		Side:        params.Side,
		Type:        params.OrderType,
		Quantity:    quantity,
		MarginAsset: marginAsset, // Use the correct margin asset from contract info
		ReduceOnly:  params.ReduceOnly,
		PositionID:  params.PositionID,
	}

	// For limit orders, round the price to the correct precision
	if (params.OrderType == "LIMIT" || params.OrderType == "STOP_LIMIT") && params.Price > 0 {
		orderParams.Price = roundToDecimal(params.Price, contractInfo.PricePrecision)
	}

	// For stop orders, set the stop price
	if (params.OrderType == "STOP_MARKET" || params.OrderType == "STOP_LIMIT") && params.StopPrice > 0 {
		orderParams.StopPrice = roundToDecimal(params.StopPrice, contractInfo.PricePrecision)
	}

	log.Default().Printf("Placing order with params: %+v\n", orderParams)

	// Place the order using the standard PlaceOrder method
	responseMap, err := api.PlaceOrder(orderParams)
	if err != nil {
		return nil, err
	}

	// Convert the map response to a structured OrderResponse
	var orderResponse OrderResponse

	// Convert the map to JSON
	jsonData, err := json.Marshal(responseMap)
	if err != nil {
		return nil, fmt.Errorf("error marshaling response: %v", err)
	}

	// Parse JSON into OrderResponse struct
	if err := json.Unmarshal(jsonData, &orderResponse); err != nil {
		return nil, fmt.Errorf("error parsing response into OrderResponse: %v", err)
	}

	return &orderResponse, nil
}
func (api *OrderAPI) BulletMap(params BulletParams) (OrderResponse, error) {
	// Get contract info for the symbol
	contractInfo, ok := api.client.ExchangeInfo[params.Symbol]
	if !ok {
		return OrderResponse{}, fmt.Errorf("symbol %s not found in exchange info", params.Symbol)
	}

	// Validate order type
	validOrderTypes := []OrderType{"MARKET", "LIMIT", "STOP_MARKET", "STOP_LIMIT"}
	isValidType := false
	for _, orderType := range validOrderTypes {
		if params.OrderType == orderType {
			isValidType = true
			break
		}
	}

	if !isValidType {
		return OrderResponse{}, fmt.Errorf("invalid order type: %s. Must be MARKET, LIMIT, STOP_MARKET, or STOP_LIMIT", params.OrderType)
	}

	// Validate required parameters for specific order types
	if (params.OrderType == "LIMIT" || params.OrderType == "STOP_LIMIT") && params.Price <= 0 {
		return OrderResponse{}, fmt.Errorf("price must be specified and greater than 0 for %s orders", params.OrderType)
	}

	if (params.OrderType == "STOP_MARKET" || params.OrderType == "STOP_LIMIT") && params.StopPrice <= 0 {
		return OrderResponse{}, fmt.Errorf("stopPrice must be specified and greater than 0 for %s orders", params.OrderType)
	}

	// Determine the minimum quantity based on order type
	var minQuantity float64
	var maxQuantity float64

	if params.OrderType == "MARKET" || params.OrderType == "STOP_MARKET" {
		minQuantity = contractInfo.MarketMinQuantity
		maxQuantity = contractInfo.MarketMaxQuantity
	} else {
		minQuantity = contractInfo.MinQuantity
		maxQuantity = contractInfo.MaxQuantity
	}

	// If min quantity is not set (could happen if filter parsing failed), use a safe default
	if minQuantity <= 0 {
		log.Default().Printf("Warning: Minimum quantity for %s not set, using default value\n", params.Symbol)
		minQuantity = 0.001 // Default fallback
	}

	// Calculate quantity based on minimum quantity and count
	quantity := minQuantity * params.Count

	// Check if quantity exceeds the maximum
	if maxQuantity > 0 && quantity > maxQuantity {
		return OrderResponse{}, fmt.Errorf("calculated quantity %.8f exceeds maximum allowed %.8f for %s",
			quantity, maxQuantity, params.Symbol)
	}

	// Round to the correct precision
	quantity = roundToDecimal(quantity, contractInfo.QuantityPrecision)

	// Check if the order type is supported for this symbol
	// For stop orders, we check if the base type (MARKET/LIMIT) is supported
	baseOrderType := params.OrderType
	if params.OrderType == "STOP_MARKET" {
		baseOrderType = "MARKET"
	} else if params.OrderType == "STOP_LIMIT" {
		baseOrderType = "LIMIT"
	}

	orderTypeSupported := false
	for _, supportedType := range contractInfo.OrderTypes {
		if supportedType == baseOrderType {
			orderTypeSupported = true
			break
		}
	}

	if !orderTypeSupported {
		return OrderResponse{}, fmt.Errorf("order type %s not supported for symbol %s",
			baseOrderType, params.Symbol)
	}

	// Determine default margin asset based on contract info
	marginAsset := contractInfo.QuoteAsset
	if len(contractInfo.MarginAssets) > 0 {
		marginAsset = contractInfo.MarginAssets[0]
	}

	// Set up order parameters
	orderParams := PlaceOrderParams{
		Symbol:      params.Symbol,
		Side:        params.Side,
		Type:        params.OrderType,
		Quantity:    quantity,
		PlaceType:   "ORDER_FORM",
		MarginAsset: marginAsset, // Use the correct margin asset from contract info
		ReduceOnly:  params.ReduceOnly,
		Leverage:    params.Leverage,
	}

	// For limit orders, round the price to the correct precision
	if (params.OrderType == "LIMIT" || params.OrderType == "STOP_LIMIT") && params.Price > 0 {
		orderParams.Price = roundToDecimal(params.Price, contractInfo.PricePrecision)
	}

	// For stop orders, set the stop price
	if (params.OrderType == "STOP_MARKET" || params.OrderType == "STOP_LIMIT") && params.StopPrice > 0 {
		orderParams.StopPrice = roundToDecimal(params.StopPrice, contractInfo.PricePrecision)
	}

	log.Default().Printf("Placing order with params: %+v\n", orderParams)

	// Place the order using the standard PlaceOrder method
	return api.PlaceOrder(orderParams)
}

// roundToDecimal rounds a float to the specified decimal places
func roundToDecimal(value float64, precision int) float64 {
	multiplier := math.Pow10(precision)
	return math.Round(value*multiplier) / multiplier
}
